src/pip/_internal/utils/setuptools_build.py

```diff
@@ -12,8 +12,8 @@
 # invoking via the shim.  This avoids e.g. the following manifest_maker
 # warning: "warning: manifest_maker: standard file '-c' not found".
 _SETUPTOOLS_SHIM = (
-    "import sys, setuptools, tokenize; sys.argv[0] = {0!r}; __file__={0!r};"
-    "f=getattr(tokenize, 'open', open)(__file__);"
+    "import sys, setuptools.build_meta, tokenize; sys.argv[0] = {0!r}; __file__={0!r};"
+    "f = setuptools.build_meta._open_setup_script(__file__) if hasattr(setuptools.build_meta, '_open_setup_script') else getattr(tokenize, 'open', open)(__file__);"
```

Comment:
I'm not particularly keen on using a private, non-standard attribute of the setuptools PEP 517 interface here. We should instead work on standardising a PEP 517 interface for editable mode, which will address this issue for *all* build backends, not just for setuptools projects without a `setup.py`.

If you want editable mode, why not just add a minimal `setup.py`? After all, if you're using editable mode, you (by definition!) have the ability to modify the project.

Reply:
Fair.

OTOH, this is the main issue for setup.cfg-only projects and I think this is a fairly straightforward solution to resolving the only blocker for making that an easy blanket recommendation (for projects where developers can expected to be using new enough pip versions) without adding much complexity if we take the approach that @uranusjr suggested in his comment.

Reply:
Yeah, I don't mind @uranusjr's approach. I *do* mind using backend internals, though. Longer term, I do think that standardising editable mode is important, and I dislike continually special-casing setuptools in ways that take the pressure off doing the broader fix. But practicality beats purity here.